### PR TITLE
Added support for butterknife library for view bindings

### PR DESCRIPTION
**Pull request by maskaravivek**
As per the discussion in #395 have added support for Butterknife. Have refactored a few existing fragments to start using Butterknife. Have tested the existing functionality and it seems to be working as expected. Will come up with the wiki doc for using it in the codebase by this weekend.

### CODE REVIEW

**Comment by misaochan** [comment]
Hrmm. Codacy does not appear to be very happy with this, although it could be an issue with how our codacy is set up? AFAIK we can customize it to work with butterknife (and probably should, if we want to proceed with this).

**Comment by maskaravivek** [comment]
@misaochan Yes I was checking with other developers to be sure that this is a build box integration setting issue. Can we customize it to suppress these warnings.

**Comment by whym** [comment]
It would be good if it is disabled only for @BindView (and other related annotations, if any).  It is a good check otherwise, for example against cases of simply (and unintentionally) forgetting to add "private".

**Comment by maskaravivek** [comment]
I agree with @whym. @misaochan can you take up the change in codacy settings. I can continue refactoring the rest of the view bindings after this gets merged.

**Comment by whym** [comment]
Isn't the configuration contained in checkstyle.xml under scripts/? Haven't tried but "ignoreAnnotationCanonicalNames" mentioned in http://checkstyle.sourceforge.net/config_design.html#VisibilityModifier seems relevant.

I think the configuration change should be added into this commit.

**Comment by maskaravivek** [comment]
Can someone help me with it. I have added the `checkstyle.xml` settings as well. It still isn't passing the codacy tests. @misaochan @whym @tobias47n9e

**Comment by misaochan** [comment]
I have no idea. :/ @nicolasraoul or @veyndan maybe?

**Comment by misaochan** [comment]
I have no idea. :/ @nicolas-raoul or @veyndan maybe?

**Comment by whym** [comment]
Adding

> 
>             <property name="packageAllowed" value="false"/>
>             <property name="protectedAllowed" value="true"/>

seems to work, at least for my local invocation of "gradle check".

**Comment by maskaravivek** [comment]
Yes @whym this didnt help. Can we try and figure out how to resolve the codacy issue.

**Comment by whym** [comment]
Does it happen only on Codacy? Or does it happen both on Codacy and locally? Sharing a copy of your local result of checkstyle might help. (It should be found at app/build/reports/checkstyle/checkstyle.xml.)

File: script/style/checkstyle.xml [hunk 1]
```diff
@@ -79,6 +79,8 @@
         </module>
 
         <module name="VisibilityModifier">
+            <property name="packageAllowed" value="true"/>
```
Comment (whym) [inline script/style/checkstyle.xml]: Did you try setting packageAllowed to false, not true? On my machine it somehow worked.

**Comment by whym** [comment]
I temporarily changed the Codacy settings to disable "Pull Request Comment" (which I believe sends these dozens of inline comments), to keep this page readable.

**Comment by maskaravivek** [comment]
@whym How do i run checkstyle task locally? I am using 
```
./gradlew checkstyle
```

Is this how i am supposed to run it locally. I dont see any scoping issues when i run it locally. :(

**Comment by maskaravivek** [comment]
Did some magic happen 💃 . It is passing the checks now. :)

**Comment by whym** [comment]
I found that PMD was enabled on Codacy, in addition to Checkstyle. PMD is now disabled (at least set to disabled, no idea why I see some of the issues still are linked to PMD...) on Codacy.

By the way, you can run checkstyle by:
     ./gradlew :app:checkstyle

**Comment by maskaravivek** [comment]
@whym does this patch require any more changes from my side before a review?

**Comment by misaochan** [comment]
The checks are passing now, I think it should be merge-able after the conflicts are resolved?

**Comment by maskaravivek** [comment]
@misaochan Rebased.

**Comment by misaochan** [comment]
Thanks @maskaravivek , I'm just waiting on @whym 's review.

**Comment by whym** [comment]
Can [the five TextViews in AboutActivity](https://github.com/commons-app/apps-android-commons/pull/409/files#diff-498c37e07962883938564ea28057ebc2) be an application of ["the view binding aspect of data binding"](https://github.com/commons-app/apps-android-commons/issues/424#issuecomment-285877465) as mentioned by @veyndan?

To be clear, I just wanted know more precisely about the possibility of what could be replaced with data binding - i.e. overlaps between butter knife and data binding. I don't meant suggest we immediately use data binding for these.

**Comment by veyndan** [comment]
Yes that could be replaced with data binding. All that you'd need to do is replace `setContentView(R.layout.activity_about);` with `ActivityAboutBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_about);` where `ActivityAboutBinding` is generated by the data binding library when the whole of the `R.layout.activity_about` layout is wrapped in a `<layout>…</layout>`.

Now referencing the views is as easy as doing `binding.aboutVersion` for `R.id.about_version`, `binding.aboutLicense` for `R.id.about_license`, and so on.

**Comment by maskaravivek** [comment]
@whym Is this fine to be merged now? :)

**Comment by whym** [comment]
To be honest, I'm not sure about it. Do we decide that we are going to prefer butter knife over data binding in places where both can be used? Or do we want to let individual contributors choose whichever they like in each patch?

**Comment by veyndan** [comment]
ButterKnife can always be used where the view binding aspect of data binding is used. I don't know why you'd ever prefer ButterKnife over data binding though with all the benefits that I have mentioned previously, and virtually no benefits of using ButterKnife over data binding.

As I've mentioned before, anywhere `@BindView(…)` is used, this can be replaced with data binding. Any of the other annotations provided by ButterKnife cannot be replaced with the view binding aspect of data binding (e.g. `@BindColor(…)`, @OnClick(…)`).

**Comment by misaochan** [comment]
I would support individual contributors choosing whichever they like (or none at all) in each patch. Not very uniform, so it wouldn't be a permanent solution, but would probably be better for the time being while we all get used to the new frameworks?

**Comment by whym** [comment]
Sounds good. As long as we recognize the mixed situation and respect each other, I'm fine with that.

**Comment by misaochan** [comment]
@whym yeah, I agree. We all good to merge this, then?

**Comment by whym** [comment]
It looks like we are good to go. @maskaravivek do you want to update the commits into a single one, as you mentioned earlier?

**Comment by whym** [comment]
... or maybe we can simply use the GitHub's "Squash and merge" feature.

**Comment by misaochan** [comment]
Oh, this conflicts with the theme toggle PR. :/ Sorry that we had to put you through resolving so many conflicts @maskaravivek , we should really be able to merge after this one.

BTW would greatly appreciate the Butterknife wiki for new contributors that you said you would make. :) Are you able to create new Wiki pages here? If you can't, tell me and I'll create one for you.

**Comment by maskaravivek** [comment]
@whym I have resolved the merge conflicts and squashed my commits to a single commit. 

@misaochan Yes i am able to create new wiki pages. Have created a [WIP page](https://github.com/commons-app/apps-android-commons/wiki/Code-Style) for butterknife wiki. I will update it with usage instructions.

**Pull request merged**